# Introduction
[Snakemake](https://snakemake.readthedocs.io/en/stable/) was used in an attempt to trim, align, perform differential expression analysis, and ultimately quality check the final results. The experience was a bit rough.

Initially, the [rna-seq-star-deseq2](https://github.com/snakemake-workflows/rna-seq-star-deseq2) and [dna-seq-gatk-variant-calling](https://github.com/snakemake-workflows/dna-seq-gatk-variant-calling) workflow pipelines were consulted. The objective was to take several Mesembryanthemum crystalinum sample sets, collected at different times and under control and stress conditions, and perform the differential expression analysis. Ultimately this was successful, however there were some workarounds necessary to effectively utilize snakemake. The biggest noted limitation was that the workflow had steps where the input and output were the same file, or were shared between steps. For example, an output BAM file needed to be indexed before a search could be performed. There are several different ways that this could be handled in snakemake, but they're not terrifically intuitive.

This repository contains the framework config and scripts at the end of the last run. Everything begins from the Snakefile. The HPC cluster configuration was used for parts of the run, consult the snakemake documentation for full usage details. The script wrapper.sh mostly contains the command-line invocation of snakemake.
